### PR TITLE
Error parsing

### PR DESCRIPTION
**Pull request by briancohan**
refactored parsing module to be more modular and maintainable. added tests to ensure expected functionality.

### CODE REVIEW

File: pytui/parser.py [hunk 1]
```diff
@@ -1,29 +1,47 @@
...
-    """
-    Extract error message and packages involved in stack trace.
 
-    The specific error message is assumed to be the last line that is not indented.
-    """
-    error_msg = ''
+def _error_message(txt: str) -> str:
+    """The specific error message is assumed to be the last line that is not indented."""
+    error_message = ''
     for line in txt.splitlines():
```
Comment (VagishVela) [inline pytui/parser.py:9]: Looks like it is having an error here, when I ran `pytui tests/test_error_parsing.py`:

```
(.venv) ➜  Perceptive-Porcupines-Code-Jam-8 git:(error-parsing) pytui tests/test_error_parsing.py
/Users/xxx/.pyenv/versions/3.9.5/lib/python3.9/asyncio/base_events.py:681: ResourceWarning: unclosed event loop <_UnixSelectorEventLoop running=False closed=False debug=False>
  _warn(f"unclosed event loop {self!r}", ResourceWarning, source=self)
ResourceWarning: Enable tracemalloc to get the object allocation traceback
╭─────────────────────────────── Traceback (most recent call last) ────────────────────────────────╮
│ /Users/xxx/.pyenv/versions/3.9.5/lib/python3.9/site-packages/textual/app.py:209 in        │
│ process_messages                                                                                 │
│                                                                                                  │
│   206 │   │   │   traceback: Traceback | None = None                                             │
│   207 │   │   │   await self.animator.start()                                                    │
│   208 │   │   │   try:                                                                           │
│ ❱ 209 │   │   │   │   await super().process_messages()                                           │
│   210 │   │   │   except Exception:                                                              │
│   211 │   │   │   │   traceback = Traceback(show_locals=True)                                    │
│   212                                                                                            │
│                                                                                                  │
│ ╭─────────────────────────────── locals ────────────────────────────────╮                        │
│ │ __class__ = <class 'textual.app.App'>                                 │                        │
│ │    driver = <textual._linux_driver.LinuxDriver object at 0x101b77af0> │                        │
│ │      self = Display(title='Megasoma Application')                     │                        │
│ │ traceback = None                                                      │                        │
│ ╰───────────────────────────────────────────────────────────────────────╯                        │
│                                                                                                  │
│ /Users/xxx/.pyenv/versions/3.9.5/lib/python3.9/site-packages/textual/message_pump.py:160  │
│ in process_messages                                                                              │
│                                                                                                  │
│   157 │   │   │   │   │   break                                                                  │
│   158 │   │   │                                                                                  │
│   159 │   │   │   try:                                                                           │
│ ❱ 160 │   │   │   │   await self.dispatch_message(message)                                       │
│   161 │   │   │   except Exception as error:                                                     │
│   162 │   │   │   │   log.exception("error in dispatch_message")                                 │
│   163 │   │   │   │   raise                                                                      │
│                                                                                                  │
│ ╭─────────────────────── locals ───────────────────────╮                                         │
│ │ idle_handler = None                                  │                                         │
│ │      message = Startup()                             │                                         │
│ │      pending = None                                  │                                         │
│ │         self = Display(title='Megasoma Application') │                                         │
│ ╰──────────────────────────────────────────────────────╯                                         │
│                                                                                                  │
│ /Users/xxx/.pyenv/versions/3.9.5/lib/python3.9/site-packages/textual/message_pump.py:175  │
│ in dispatch_message                                                                              │
│                                                                                                  │
│   172 │   async def dispatch_message(self, message: Message) -> bool | None:                     │
│   173 │   │   if isinstance(message, events.Event):                                              │
│   174 │   │   │   if not isinstance(message, events.Null):                                       │
│ ❱ 175 │   │   │   │   await self.on_event(message)                                               │
│   176 │   │   else:                                                                              │
│   177 │   │   │   return await self.on_message(message)                                          │
│   178 │   │   return False                                                                       │
│                                                                                                  │
│ ╭──────────────────── locals ─────────────────────╮                                              │
│ │ message = Startup()                             │                                              │
│ │    self = Display(title='Megasoma Application') │                                              │
│ ╰─────────────────────────────────────────────────╯                                              │
│                                                                                                  │
│ /Users/xxx/.pyenv/versions/3.9.5/lib/python3.9/site-packages/textual/app.py:279 in        │
│ on_event                                                                                         │
│                                                                                                  │
│   276 │   │   │   │   await self.focused.forward_event(event)                                    │
│   277 │   │   │   await self.view.forward_event(event)                                           │
│   278 │   │   else:                                                                              │
│ ❱ 279 │   │   │   await super().on_event(event)                                                  │
│   280 │                                                                                          │
│   281 │   async def action(                                                                      │
│   282 │   │   self, action: str, default_namespace: object | None = None                         │
│                                                                                                  │
│ ╭───────────────────── locals ──────────────────────╮                                            │
│ │ __class__ = <class 'textual.app.App'>             │                                            │
│ │     event = Startup()                             │                                            │
│ │      self = Display(title='Megasoma Application') │                                            │
│ ╰───────────────────────────────────────────────────╯                                            │
│                                                                                                  │
│ /Users/xxx/.pyenv/versions/3.9.5/lib/python3.9/site-packages/textual/message_pump.py:185  │
│ in on_event                                                                                      │
│                                                                                                  │
│   182 │   │                                                                                      │
│   183 │   │   dispatch_function: MessageHandler = getattr(self, method_name, None)               │
│   184 │   │   if dispatch_function is not None:                                                  │
│ ❱ 185 │   │   │   await dispatch_function(event)                                                 │
│   186 │   │   if event.bubble and self._parent and not event._stop_propagaton:                   │
│   187 │   │   │   if event.sender == self._parent:                                               │
│   188 │   │   │   │   pass                                                                       │
│                                                                                                  │
│ ╭─────────────────────────────────────────── locals ───────────────────────────────────────────╮ │
│ │ dispatch_function = <bound method Display.on_startup of Display(title='Megasoma              │ │
│ │                     Application')>                                                           │ │
│ │             event = Startup()                                                                │ │
│ │       method_name = 'on_startup'                                                             │ │
│ │              self = Display(title='Megasoma Application')                                    │ │
│ ╰──────────────────────────────────────────────────────────────────────────────────────────────╯ │
│                                                                                                  │
│ /Users/xxx/.pyenv/versions/3.9.5/lib/python3.9/site-packages/pytui/display.py:43 in       │
│ on_startup                                                                                       │
│                                                                                                  │
│   40 │   │   """App layout"""                                                                    │
│   41 │   │   view = await self.push_view(DockView())                                             │
│   42 │   │                                                                                       │
│ ❱ 43 │   │   self.data = get_all_error_results()                                                 │
│   44 │   │                                                                                       │
│   45 │   │   header = Header(f"{APP_NAME}: {self.data['error_message']}")                        │
│   46 │   │   footer = Footer()                                                                   │
│                                                                                                  │
│ ╭─────────────────── locals ────────────────────╮                                                │
│ │ event = Startup()                             │                                                │
│ │  self = Display(title='Megasoma Application') │                                                │
│ │  view = DockView(name='DockView#1')           │                                                │
│ ╰───────────────────────────────────────────────╯                                                │
│                                                                                                  │
│ /Users/xxx/.pyenv/versions/3.9.5/lib/python3.9/site-packages/pytui/core.py:31 in          │
│ get_all_error_results                                                                            │
│                                                                                                  │
│   28 │   StackOverflow backend and all the results are returned.                                 │
│   29 │   """                                                                                     │
│   30 │   all_text = run_and_get_stderr()                                                         │
│ ❱ 31 │   parsed = parse_stacktrace(all_text)                                                     │
│   32 │                                                                                           │
│   33 │   # Copy the error to the clipboard if asked                                              │
│   34 │   if parsed["error_message"]:                                                             │
│                                                                                                  │
│ ╭────── locals ──────╮                                                                           │
│ │    all_text = None │                                                                           │
│ │ max_results = 10   │                                                                           │
│ ╰────────────────────╯                                                                           │
│                                                                                                  │
│ /Users/xxx/.pyenv/versions/3.9.5/lib/python3.9/site-packages/pytui/parser.py:39 in        │
│ parse_stacktrace                                                                                 │
│                                                                                                  │
│   36                                                                                             │
│   37 def parse_stacktrace(txt: str) -> dict[str, Any]:                                           │
│   38 │   """Extract error message and packages involved in stack trace."""                       │
│ ❱ 39 │   error_message = _error_message(txt)                                                     │
│   40 │   files = _error_files(txt)                                                               │
│   41 │   packages = _extract_packages(files)                                                     │
│   42                                                                                             │
│                                                                                                  │
│ ╭── locals ──╮                                                                                   │
│ │ txt = None │                                                                                   │
│ ╰────────────╯                                                                                   │
│                                                                                                  │
│ /Users/xxx/.pyenv/versions/3.9.5/lib/python3.9/site-packages/pytui/parser.py:9 in         │
│ _error_message                                                                                   │
│                                                                                                  │
│    6 def _error_message(txt: str) -> str:                                                        │
│    7 │   """The specific error message is assumed to be the last line that is not indented."     │
│    8 │   error_message = ''                                                                      │
│ ❱  9 │   for line in txt.splitlines():                                                           │
│   10 │   │   if line == line.strip():                                                            │
│   11 │   │   │   error_message = line.strip()                                                    │
│   12 │   return error_message                                                                    │
│                                                                                                  │
│ ╭─────── locals ───────╮                                                                         │
│ │ error_message = ''   │                                                                         │
│ │           txt = None │                                                                         │
│ ╰──────────────────────╯                                                                         │
╰──────────────────────────────────────────────────────────────────────────────────────────────────╯
AttributeError: 'NoneType' object has no attribute 'splitlines'
```

**Review by VagishVela — CHANGES_REQUESTED** [review]
Looks like `txt` of NoneType is being passes into `def _error_message` causing the failure.
Can you have a look at this again @mustafaquraish ?

File: pytui/parser.py (hunk 1 above)
Comment (briancohan) [reply pytui/parser.py:9]: Try running with `pytest`. You're not supposed to run anything in `tests` manually.

Comment (VagishVela) [reply pytui/parser.py:9]: @briancohan If it runs fine with that then merge. 😄 I ran the example and it worked as well.

Comment (VagishVela) [reply pytui/parser.py:9]: My bad, I must have been thinking from the examples point of view. Great that we are implementing actual tests. I ran the tests and all looks good! 👍🏾

**Review by VagishVela — APPROVED** [review]
LGTM

**Pull request merged**